File: src/com/edlplan/ui/fragment/BaseFragment.java
package com.edlplan.ui.fragment;

import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.edlplan.framework.easing.Easing;
import com.edlplan.ui.ActivityOverlay;
import com.edlplan.ui.EasingHelper;

import ru.nsu.ccfit.zuev.osuplus.R;

public abstract class BaseFragment extends Fragment {

    private View root;

    private View background;

    private boolean dismissOnBackgroundClick = false;

    private boolean created = false;

    public boolean isDismissOnBackgroundClick() {
        return dismissOnBackgroundClick;
    }

    public void setDismissOnBackgroundClick(boolean dissmissOnBackgroundClick) {
        this.dismissOnBackgroundClick = dissmissOnBackgroundClick;
    }

    public boolean isCreated() {
        return created;
    }

    public View getRoot() {
        return root;
    }

    public @IdRes
    int getBackgroundId() {
        return R.id.frg_background;
    }

    @SuppressWarnings("unchecked")
    public <T extends View> T findViewById(@IdRes int id) {
        Object o = getRoot() != null ? getRoot().findViewById(id) : null;
        if (o == null) {
            return null;
        } else {
            return (T) o;
        }
    }

    protected abstract @LayoutRes
    int getLayoutID();

    protected abstract void onLoadView();

    protected void playBackgroundHideInAnim(int duration) {
        View background = findViewById(R.id.frg_background);
        if (background != null) {
            background.setAlpha(0);
            background.animate().cancel();
            background.animate()
                    .alpha(1)
                    .setDuration(duration)
                    .setInterpolator(EasingHelper.asInterpolator(Easing.InOutQuad))
                    .start();
        }
    }

    protected void playBackgroundHideOutAnim(int duration) {
        View background = findViewById(R.id.frg_background);
        if (background != null) {
            background.animate().cancel();
            background.animate()
                    .alpha(0)
                    .setDuration(duration)
                    .setInterpolator(EasingHelper.asInterpolator(Easing.InOutQuad))
                    .start();
        }
    }


    public void show() {
        ActivityOverlay.addOverlay(this, this.getClass().getName() + "@" + this.hashCode());
    }

    public void dismiss() {
        ActivityOverlay.dismissOverlay(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        created = true;
        root = inflater.inflate(getLayoutID(), container, false);
        if (findViewById(getBackgroundId()) != null) {
            background = findViewById(getBackgroundId());
            background.setOnClickListener(v -> {
                if (isDismissOnBackgroundClick()) {
                    dismiss();
                }
            });
        }
        onLoadView();
        return root;
    }
}

File: src/com/edlplan/ui/fragment/BaseFragment.kt
package com.edlplan.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment
import com.edlplan.framework.easing.Easing
import com.edlplan.ui.ActivityOverlay
import com.edlplan.ui.EasingHelper
import com.reco1l.osu.*
import ru.nsu.ccfit.zuev.osuplus.R

abstract class BaseFragment : Fragment(), BackPressListener {
    var root: View? = null
        private set
    private var background: View? = null
    var onDismissListener: OnDismissListener? = null
    var isDismissOnBackgroundClick = false
    var isCreated = false
        private set
    var isDismissOnBackPress = true


    /**
     * If true, the fragment will intercept back press event when it's received.
     */
    var interceptBackPress = true


    private var isLoaded = false

    private var isDismissCalled = false


    @get:IdRes
    val backgroundId: Int
        get() = R.id.frg_background

    @Suppress("UNCHECKED_CAST")
    fun <T : View?> findViewById(@IdRes id: Int): T? {
        val o: Any? = if (root != null) root!!.findViewById<View>(id) else null
        return if (o == null) {
            null
        } else {
            o as T
        }
    }

    @get:LayoutRes
    protected abstract val layoutID: Int
    protected abstract fun onLoadView()
    protected fun playBackgroundHideInAnim(duration: Int) {
        val background = findViewById<View>(R.id.frg_background)
        if (background != null) {
            background.alpha = 0f
            background.animate().cancel()
            background.animate()
                    .alpha(1f)
                    .setDuration(duration.toLong())
                    .setInterpolator(EasingHelper.asInterpolator(Easing.InOutQuad))
                    .start()
        }
    }

    protected fun playBackgroundHideOutAnim(duration: Int) {
        val background = findViewById<View>(R.id.frg_background)
        if (background != null) {
            background.animate().cancel()
            background.animate()
                    .alpha(0f)
                    .setDuration(duration.toLong())
                    .setInterpolator(EasingHelper.asInterpolator(Easing.InOutQuad))
                    .start()
        }
    }

    open fun show() {
        ActivityOverlay.addOverlay(this, this.javaClass.name + "@" + this.hashCode())
    }

    open fun dismiss() {
        isDismissCalled = true

        if (isLoaded) {
            isDismissCalled = false
            ActivityOverlay.dismissOverlay(this)
            onDismissListener?.OnDismiss()
        }
    }

    fun save() {
        ActivityOverlay.saveState(this.javaClass.name + "@" + this.hashCode())
    }

    fun load(): SavedState? {
        return ActivityOverlay.loadState(this.javaClass.name + "@" + this.hashCode())
    }

    override fun callDismissOnBackPress() {
        if (isDismissOnBackPress) {
            dismiss()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        isLoaded = false
        isCreated = true
        root = inflater.inflate(layoutID, container, false)
        findViewById<View>(backgroundId)?.setOnClickListener {
            if (isDismissOnBackgroundClick) {
                dismiss()
            }
        }

        onLoadView()
        isLoaded = true

        if (isDismissCalled) {
            dismiss()
            isDismissCalled = false
        }

        return root
    }

    fun interface OnDismissListener {
        fun OnDismiss()
    }
}